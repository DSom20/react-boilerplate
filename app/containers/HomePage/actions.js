/*
 *
 * HomePage actions
 *
 */

import {
  FETCH_TIDBITS,
  FETCH_TIDBITS_SUCCESS,
  FETCH_TIDBITS_FAILURE,
} from './constants';

export function fetchTidbits() {
  return {
    type: FETCH_TIDBITS,
  };
}

export function fetchTidbitsSuccess(tidbits) {
  return {
    type: FETCH_TIDBITS_SUCCESS,
    tidbits,
  };
}

export function fetchTidbitsFailure(error) {
  return {
    type: FETCH_TIDBITS_FAILURE,
    error,
  };
}
