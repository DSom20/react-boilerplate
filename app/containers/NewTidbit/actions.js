/*
 *
 * NewTidbit actions
 *
 */

import {
  CHANGE_TIDBIT,
  SAVE_TIDBIT,
  SAVE_TIDBIT_SUCCESS,
  SAVE_TIDBIT_FAILURE,
  RESET_TIDBIT,
} from './constants';

export function changeTidbit(tidbit) {
  return {
    type: CHANGE_TIDBIT,
    tidbit,
  };
}

export function saveTidbit() {
  return {
    type: SAVE_TIDBIT,
  };
}

export function saveTidbitSuccess(response) {
  return {
    type: SAVE_TIDBIT_SUCCESS,
    response,
  };
}

export function saveTidbitFailure(error) {
  return {
    type: SAVE_TIDBIT_FAILURE,
    error,
  };
}

export function resetTidbit() {
  return {
    type: RESET_TIDBIT,
  };
}

// export function defaultAction() {
//   return {
//     type: DEFAULT_ACTION,
//   };
// }
