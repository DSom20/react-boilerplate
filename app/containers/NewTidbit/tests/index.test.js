/**
 *
 * Tests for NewTidbit
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import { NewTidbit, mapDispatchToProps } from '../index';
import { changeTidbit, saveTidbit, resetTidbit } from '../actions';
// import { DEFAULT_LOCALE } from '../../../i18n';
// import { changeUsername } from '../actions';
// import { loadRepos } from '../../App/actions';
import configureStore from '../../../configureStore';

describe('<NewTidbit />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    // const dispatch = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewTidbit
            tidbit=""
            saving={false}
            resulted={false}
            error={false}
            onChangeTidbit={() => {}}
            onSubmitForm={() => {}}
            onUnmount={() => {}}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewTidbit
            tidbit=""
            saving={false}
            resulted={false}
            error={false}
            onChangeTidbit={() => {}}
            onSubmitForm={() => {}}
            onUnmount={() => {}}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('Should call onChangeTidbit with change event on input field', () => {
    const onChangeTidbitSpy = jest.fn();
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewTidbit
            tidbit=""
            saving={false}
            resulted={false}
            error={false}
            onChangeTidbit={onChangeTidbitSpy}
            onSubmitForm={() => {}}
            onUnmount={() => {}}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(onChangeTidbitSpy).not.toHaveBeenCalled();
    const inputField = getByPlaceholderText(
      'Whatever tidbit you can think of!',
    );
    fireEvent.change(inputField, { target: { value: 'a' } });
    expect(onChangeTidbitSpy).toHaveBeenCalled();
  });

  it('Should call onSubmitForm with submit event from form', () => {
    const submitSpy = jest.fn();
    const { getByTestId } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewTidbit
            tidbit=""
            saving={false}
            resulted={false}
            error={false}
            onChangeTidbit={() => {}}
            onSubmitForm={submitSpy}
            onUnmount={() => {}}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
    const form = getByTestId('tidbitForm');
    fireEvent.submit(form);
    expect(submitSpy).toHaveBeenCalled();
  });

  it('Should call onUnmount when unmounts', () => {
    const onUnmountSpy = jest.fn();
    const { unmount } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <NewTidbit
            tidbit=""
            saving={false}
            resulted={false}
            error={false}
            onChangeTidbit={() => {}}
            onSubmitForm={() => {}}
            onUnmount={onUnmountSpy}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(onUnmountSpy).not.toHaveBeenCalled();
    unmount();
    expect(onUnmountSpy).toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onChangeTidbit', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onChangeTidbit).toBeDefined();
      });

      it('should dispatch changeTidbit when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const tidbit = 'sweet tidbit!';
        const mockEvent = { target: { value: tidbit } };
        result.onChangeTidbit(mockEvent);
        expect(dispatch).toHaveBeenCalledWith(changeTidbit(tidbit));
      });
    });

    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch saveTidbit when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onSubmitForm();
        expect(dispatch).toHaveBeenCalledWith(saveTidbit());
      });

      it('should preventDefault if called with event', () => {
        const preventDefault = jest.fn();
        const result = mapDispatchToProps(() => {});
        const evt = { preventDefault };
        result.onSubmitForm(evt);
        expect(preventDefault).toHaveBeenCalled();
      });
    });

    describe('onUnmount', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onUnmount).toBeDefined();
      });

      it('should dispatch resetTidbit when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onUnmount();
        expect(dispatch).toHaveBeenCalledWith(resetTidbit());
      });
    });
  });
});
