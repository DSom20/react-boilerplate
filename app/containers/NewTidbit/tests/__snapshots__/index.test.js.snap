// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTidbit /> Should render and match the snapshot 1`] = `
<article>
  <h1>
    <span>
      Enter a new tidbit on this page!
    </span>
  </h1>
  <section>
    <form
      class="Form-sc-66efnp-0 kwWrFk"
      data-testid="tidbitForm"
    >
      <label
        for="tidbit"
      >
        <span>
          Enter Tidbit: 
        </span>
        <input
          class="Input-fnidfs-0 bXmxGP"
          id="tidbit"
          placeholder="Whatever tidbit you can think of!"
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Submit Tidbit!
      </button>
    </form>
    <div />
  </section>
</article>
`;
